src/common/io_exerciser/DataGenerator.cc

```diff
@@ -0,0 +1,713 @@
+}
+
+const DataGenerator::DataGenerationSingleton& 
+DataGenerator::DataGenerationSingleton::createSpecificInstance(uint64_t unique_id)
+{
+  std::scoped_lock lock(m_mutex);
+  ceph_assert(!m_singletonInstance.m_created);
+  m_singletonInstance = DataGenerationSingleton(unique_id);
+
+  return m_singletonInstance;
+}
+
+const DataGenerator::DataGenerationSingleton&
+  DataGenerator::DataGenerationSingleton::getInstance()
+{
+  if (!m_singletonInstance.m_created)
+  {
+    std::scoped_lock lock(m_mutex);
+    if (!m_singletonInstance.m_created)
+    {
+      std::mt19937_64 random_generator = 
+        std::mt19937_64(duration_cast<std::chrono::milliseconds>(
+          std::chrono::system_clock::now().time_since_epoch()).count());
+      m_singletonInstance = DataGenerator::DataGenerationSingleton(random_generator());
+    }
+  }
+  return m_singletonInstance;
+}
+
+const uint64_t DataGenerator::DataGenerationSingleton::getUniqueId()
+{
+  return getInstance().m_uniqueId;
+}
+
+void DataGenerator::generate_wrong_data(uint64_t offset, uint64_t length,
+                                        bufferlist& retlist)
+{
+  uint64_t block_size = m_model.get_block_size();
+  char buffer[block_size];
+  for (uint64_t block_offset = offset;
+       block_offset < offset + length;
+       block_offset++)
+  { 
+    std::memset(buffer, 0, block_size);
+    retlist.append(ceph::bufferptr(buffer, block_size));
+  }
+}
+
+bool DataGenerator::validate(bufferlist& bufferlist, uint64_t offset, uint64_t length)
+{
+  ceph::bufferlist comparison_list;
+  generate_data(offset, length, comparison_list);
+  return bufferlist.contents_equal(comparison_list);
+}
+
+#include <bitset>
+
+ceph::bufferptr SeededRandomGenerator::generate_block(uint64_t block_offset)
+{
+  uint64_t block_size = m_model.get_block_size();
+  char buffer[block_size];
+  
+  std::mt19937_64 random_generator(m_model.get_seed(block_offset));
+  uint64_t rand1 = random_generator();
+  uint64_t rand2 = random_generator();
+
+  constexpr size_t generation_length = sizeof(uint64_t);
+
+  for (uint64_t i = 0; i < block_size; i+=(2*generation_length), rand1++, rand2--)
+  {
+    std::memcpy(buffer + i, &rand1, generation_length);
+    std::memcpy(buffer + i + generation_length, &rand2, generation_length);
+  }
+
+  size_t remainingBytes = block_size % (generation_length * 2);
+  if (remainingBytes > generation_length)
+  {
+    size_t remainingBytes2 = remainingBytes - generation_length;
+    std::memcpy(buffer + block_size - remainingBytes, &rand1, remainingBytes);
+    std::memcpy(buffer + block_size - remainingBytes2, &rand2, remainingBytes2);
+  }
+  else if (remainingBytes > 0)
+  {
+    std::memcpy(buffer + block_size - remainingBytes, &rand1, remainingBytes);
+  }
+
+  return ceph::bufferptr(buffer, block_size);
+}
+
+void SeededRandomGenerator::generate_data(uint64_t offset, uint64_t length,
+                                          bufferlist& retlist)
+{
+  ceph_assert(retlist.length() == 0);
```

Comment:
so - if you are sure that 'relist' is out-only - it would probably be better
to change the signature to:

bufferlist generate...(uint64_t offset, uint64_t length)

Reply:
Not sure why I did it like this originally - have updated.